python/sglang/srt/managers/cache_controller.py

```diff
@@ -659,13 +659,14 @@ def _generic_page_get(self, operation, hash_values, host_indices):
                     f"Prefetch operation {operation.request_id} failed to retrieve page {hash_values[i]}."
                 )
                 break
-            if operation.increment(self.page_size):
-                self.mem_pool_host.set_from_flat_data_page(
-                    host_indices[i * self.page_size],
-                    page_data[i],
-                )
-            else:
-                break
+            # Must set the data before increasing the completed tokens.
```

Comment:
<img width="710" height="494" alt="image" src="https://github.com/user-attachments/assets/ea225d78-a916-4226-807a-6fc2aa8e81ae" />

If `operation.increment(self.page_size)` is called before `mark_done()`, and after it succeeds, another thread calls `mark_done()`, while `set_from_flat_data_page` has not yet completed, there could indeed be a risk of reading dirty data.